### PR TITLE
Moiv0.9

### PR DESCRIPTION
**Pull request by jd-lara**
This is pretty much work in progress to update the solver. See https://github.com/JuliaOpt/MathOptInterface.jl/issues/736

### CODE REVIEW

**Comment by jd-lara** [comment]
@odow what's the new implementation of `LQOI.VecLin`?
```
    (LQOI.VecLin, MOI.Nonnegatives),
    (LQOI.VecLin, MOI.Nonpositives),
    (LQOI.VecLin, MOI.Zeros)
```

**Comment by odow** [comment]
They are removed. Use the full_bridge_optimizer instead.

**Comment by jd-lara** [comment]
Thanks, I haven't get there yet. The current changes are at Line 1361 of MOIWrapper

**Comment by blegat** [comment]
Call the file `MOI_wrapper`, not `MOIWrapper`

File: src/MOIWrapper.jl [hunk 1]
```diff
@@ -1,560 +1,2373 @@
...
+
+MOI.supports(::Optimizer, ::MOI.VariableName, ::Type{MOI.VariableIndex}) = true
+MOI.supports(::Optimizer, ::MOI.ConstraintName, ::Type{<:MOI.ConstraintIndex}) = true
+MOI.supports(::Optimizer, ::MOI.ObjectiveFunctionType) = true
+
+MOI.supports(::Optimizer, ::MOI.Name) = true
+MOI.supports(::Optimizer, ::MOI.Silent) = true
+MOI.supports(::Optimizer, ::MOI.TimeLimitSec) = true
+MOI.supports(::Optimizer, ::MOI.ObjectiveSense) = true
+MOI.supports(::Optimizer, ::MOI.ListOfConstraintIndices) = true
```
Comment (blegat) [inline src/MOIWrapper.jl]: See https://github.com/JuliaOpt/Gurobi.jl/pull/234

File: Project.toml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+[deps]
+Libdl = "8f399da3-3557-5675-b5ff-fb832c97cbdb"
+LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
+MathOptInterface = "b8f27783-ece8-5eb3-8dc8-9495eed66fee"
+MathProgBase = "fdba3010-5040-5b88-9595-932c9decdf73"
+OrderedCollections = "bac558e1-5e72-5ebc-8fee-abe8a469f55d"
+SparseArrays = "2f01184e-e22b-5df5-ae63-d93ebab69eaf"
+
+[compat]
+MathOptInterface = "~0.9.0"
```
Comment (blegat) [inline Project.toml]: Use v0.9.1 as v0.9.0 does not work on Julia v1.2

File: Manifest.toml [hunk 3]
```diff
@@ -0,0 +1,130 @@
+# This file is machine-generated - editing it directly is not advised
```
Comment (frapac) [inline Manifest.toml]: Do we have to version the `Manifest.toml` file?

File: src/MOI_wrapper.jl [hunk 4]
```diff
@@ -0,0 +1,2708 @@
...
+}
+
+mutable struct VariableInfo
+    index::MOI.VariableIndex
+    column::Int
+    bound::BoundType
+    type::VariableType
+    start::Union{Float64, Nothing}
+    name::String
+    # Storage for constraint names associated with variables because Gurobi
```
Comment (frapac) [inline src/MOI_wrapper.jl]: Should we replace Gurobi by Xpress here?

File: src/MOI_wrapper.jl [hunk 5]
```diff
@@ -0,0 +1,2708 @@
...
+function MOI.get(model::Optimizer, ::MOI.ListOfConstraintAttributesSet)
+    return MOI.AbstractConstraintAttribute[MOI.ConstraintName()]
+end
+
+function _indices_and_coefficients(
+    indices::AbstractVector{Int}, coefficients::AbstractVector{Float64},
+    model::Optimizer, f::MOI.ScalarAffineFunction{Float64}
+)
+    i = 1
+    for term in f.terms
```
Comment (frapac) [inline src/MOI_wrapper.jl]: Could we replace this line by this one?
`for (i, term) in enumerate(f.terms)`

File: src/MOI_wrapper.jl [hunk 6]
```diff
@@ -0,0 +1,2708 @@
...
+    I::AbstractVector{Int}, J::AbstractVector{Int}, V::AbstractVector{Float64},
+    indices::AbstractVector{Int}, coefficients::AbstractVector{Float64},
+    model::Optimizer, f::MOI.ScalarQuadraticFunction
+)
+    for (i, term) in enumerate(f.quadratic_terms)
+        I[i] = _info(model, term.variable_index_1).column
+        J[i] = _info(model, term.variable_index_2).column
+        V[i] =  term.coefficient
+        # Xpress does the same according to page 14 of the reference Manual.
+        # Gurobi returns a list of terms. MOI requires 0.5 x' Q x. So, to get
```
Comment (frapac) [inline src/MOI_wrapper.jl]: Replace Gurobi by Xpress?

File: src/MOI_wrapper.jl [hunk 7]
```diff
@@ -0,0 +1,2708 @@
...
+    optimize(model.inner)
+    model.has_infeasibility_cert =
+    MOI.get(model, MOI.DualStatus()) == MOI.INFEASIBILITY_CERTIFICATE
+    model.has_unbounded_ray =
+        MOI.get(model, MOI.PrimalStatus()) == MOI.INFEASIBILITY_CERTIFICATE
+    return
+end
+
+# These strings are taken directly from the following page of the online Gurobi
+# documentation: https://www.com/documentation/8.1/refman/optimization_status_codes.html#sec:StatusCodes
```
Comment (frapac) [inline src/MOI_wrapper.jl]: This URL seems invalid.

File: src/MOI_wrapper.jl [hunk 8]
```diff
@@ -0,0 +1,2708 @@
...
+        [model.variable_info[CleverDicts.LinearIndex(i)].index for i in cols]
+    )
+end
+
+###
+### Optimize methods.
+###
+
+function MOI.optimize!(model::Optimizer)
+    # Note: although Gurobi will call update regardless, we do it now so that
```
Comment (frapac) [inline src/MOI_wrapper.jl]: Gurobi -> Xpress?

File: src/MOI_wrapper.jl [hunk 9]
```diff
@@ -0,0 +1,2708 @@
...
+    # the appropriate `needs_update` flag is set.
+    optimize(model.inner)
+    model.has_infeasibility_cert =
+    MOI.get(model, MOI.DualStatus()) == MOI.INFEASIBILITY_CERTIFICATE
```
Comment (frapac) [inline src/MOI_wrapper.jl]: Add a tab here to improve readibility?

File: src/MOI_wrapper.jl [hunk 10]
```diff
@@ -0,0 +1,2708 @@
...
+            rethrow(ex)
+        end
+    end
+end
+
+function MOI.get(model::Optimizer, ::MOI.VariablePrimal, x::MOI.VariableIndex)
+    if model.has_unbounded_ray
+        return get_dblattrelement(model.inner, "UnbdRay", _info(model, x).column)
+    else
+        return get_dblattrelement(model.inner, "X", _info(model, x).col
```
Comment (frapac) [inline src/MOI_wrapper.jl]: Gurobi -> Xpress?

Comment (frapac) [inline src/MOI_wrapper.jl]: Do you think it could be worthwile to add a bridge `SOCToQuadBridge` in MOI?

File: Manifest.toml (hunk 3 above)
Comment (jd-lara) [reply Manifest.toml]: I will deleted when the PR is ready to merge. Makes it easier to develop in an environment.

File: src/MOI_wrapper.jl (hunk 8 above)
Comment (jd-lara) [reply src/MOI_wrapper.jl]: Haven't got this far currently still at line 1718

File: src/MOI_wrapper.jl (hunk 9 above)
Comment (jd-lara) [reply src/MOI_wrapper.jl]: Haven't got this far currently still at line 1718

File: src/MOI_wrapper.jl (hunk 10 above)
Comment (jd-lara) [reply src/MOI_wrapper.jl]: Haven't got this far currently still at line 1718

Comment (jd-lara) [reply src/MOI_wrapper.jl]: Not sure, Xpress supports both SOC and Quad and we could get to a point of supporting both

File: src/MOI_wrapper.jl (hunk 7 above)
Comment (jd-lara) [reply src/MOI_wrapper.jl]: Haven't got this far currently still at line 1718

**Comment by joaquimg** [comment]
Any plans for updates here?

**Comment by jd-lara** [comment]
@joaquimg I made some progress, I pretty much ended up copying and pasting the code from the old wrapper for the Termination Status and found the equivalent in three different places for the raw strings. Please take a look, If I got the logic right from the old wrapper then the stop status is only used when specific termination status are triggered. 

I am sort of stuck implementing the get_primal and get_dual functions starting in line 1922. My impression is that in Gurobi the values can be queried by individual variable and constraint (@odow?). However, in the old MOI wrapper (both in Gurobi and Xpress) the strategy was to return all of the primals into a container (`get_linear_primal_solution!(m, x::Vector{Float64})`) and Xpress doesn't seem to have the capability to get an attribute for a specific row and column as Gurobi has. Any pointers on how to proceed in this case? My first inclination would be to return all the solutions into a vector and then index into it. But, this will allocate a vector of all the solutions every time a specific variable or constraint gets queried.

**Comment by jd-lara** [comment]
Last line modified 2022

File: src/MOI_wrapper.jl (hunk 4 above)
Comment (jd-lara) [reply src/MOI_wrapper.jl]: Pending to check if it applied to Xpress. @joaquimg do we need this too?

**Comment by joaquimg** [comment]
Lets get the entire solution vector from Xpress right after solve and then access it.
This is pretty reasonable and probably even more efficient.

File: src/MOI_wrapper.jl (hunk 4 above)
Comment (joaquimg) [reply src/MOI_wrapper.jl]: Yes we need. Variable bound are not proper constraints in xpress also.

File: src/MOI_wrapper.jl [hunk 11]
```diff
@@ -0,0 +1,2789 @@
...
+end
+
+function MOI.add_constraints(
+    model::Optimizer, f::Vector{MOI.ScalarAffineFunction{Float64}},
+    s::Vector{<:Union{MOI.GreaterThan{Float64}, MOI.LessThan{Float64}, MOI.EqualTo{Float64}}}
+)
+    if length(f) != length(s)
+        error("Number of functions does not equal number of sets.")
+    end
+    canonicalized_functions = MOI.Utilities.canonical.(f)
```
Comment (joaquimg) [inline src/MOI_wrapper.jl]: Not sure if we need to cannonicalize. Might worth check in the manual

File: src/MOI_wrapper.jl [hunk 12]
```diff
@@ -0,0 +1,2789 @@
...
+    # First pass: compute number of non-zeros to allocate space.
+    nnz = 0
+    for fi in canonicalized_functions
+        if !iszero(fi.constant)
+            throw(MOI.ScalarFunctionConstantNotZero{Float64, eltype(f), eltype(s)}(fi.constant))
+        end
+        nnz += length(fi.terms)
+    end
```
Comment (joaquimg) [inline src/MOI_wrapper.jl]: Adding zeros might be faster than filtering

**Comment by odow** [comment]
> @odow should I copy the tests from Gurobi.jl as well and go from there?

Yes. The Gurobi tests, especially the `MOI.Test` ones are pretty comprehensive.

**Comment by jd-lara** [comment]
> Lets get the entire solution vector from Xpress right after solve and then access it.
> This is pretty reasonable and probably even more efficient.

My thinking about this is then to create a new field in model (previously names instance) struct with the complete solution and then query from there.

**Comment by joaquimg** [comment]
Vectors in the optimizer are ok. Struct with vector inside optimizer is also ok.

At some point I plan to only grab part of the solution. For instance, not querying slacks e not even duals, only primals. Then maybe do that in que lazy way. Just query the primal vector when the first primal is asked for. (None os this is needed now, just a heads up on future plans)

**Comment by joaquimg** [comment]
Let me know if you have more questions

**Comment by jd-lara** [comment]
I need to experiment more with the return objects of the get solution from Xpress. Also, the PSCC deadline is coming. I will get back to this after that.

**Comment by joaquimg** [comment]
Nice! I did imagine you could be distracted by PSCC, hope we can meet there next year ;)

File: src/MOIWrapper.jl [hunk 13]
```diff
@@ -1,560 +1,2373 @@
...
+function MOI.modify(
+    model::Optimizer,
+    ::MOI.ObjectiveFunction{MOI.ScalarAffineFunction{Float64}},
+    chg::MOI.ScalarConstantChange{Float64}
+)
+    set_dblattr!(model.inner, "ObjCon", chg.new_constant)
+
+    return
+end
+=#
```
[the remark this answers is not in the document]
Comment (jd-lara) [reply src/MOIWrapper.jl]: @joaquimg can you confirm here please?

**Comment by jd-lara** [comment]
@joaquimg is this needed in Xpress? I commented out for now to look for the Xpress equivalent of these settings. 

```
    Note that we set the parameter `InfUnbdInfo` to `1` rather than the default
    of `0` so that we can query infeasibility certificates. Users are, however,
    free to over-ride this as follows `Optimizer(InfUndbInfo=0)`. In addition,
    we also set `QCPDual` to `1` to enable duals in QCPs. Users can override
    this by passing `Optimizer(QCPDual=0)`.
```

**Comment by odow** [comment]
>  is this needed in Xpress?

I think it's just a Gurobi specific thing

**Comment by joaquimg** [comment]
Exactly, that is a Gurobi thing.
The related xpress detail is that the presolve must be disabled if you want to query infeasibility certificates. I to that explicitly on MOI_Wrapper tests

File: src/xprs_env.jl [hunk 14]
```diff
@@ -5,16 +5,23 @@
...
 Need to do ir once and only once.
 """
-function Env()
-    ret = @xprs_ccall(init, Cint, (Ptr{Cchar},), C_NULL)
-    if ret != 0
-        if ret == :number
-            error("Invalid Xpress license")
-        else
-            error("Failed to create environment (error $ret).")
+mutable struct Env
```
Comment (joaquimg) [inline src/xprs_env.jl]: not sure we should keep this...
it does not hold info like gurobi's, its just a boot in the library

**Comment by jd-lara** [comment]
closed in favor of #43

**Pull request closed without merging**